[NCN-440] Update validation for one-click access

jsperhac opened the pull request:
#### Dependencies

This work is part of Epic [NCN-434](https://sdx-sdsc.atlassian.net/browse/NCN-434), whose PRs are as follows:
- #1693, NCN-633 `plgUserHubzero` (on user login, ensure user has a secret in the DB)
- #1683 NCN-439 `com_members` (manage user secret)
- #1676  NCN-437 `com_newsletter` (manage campaign secret)
- #1675 NCN-438 `com_config` (manage hub secret)
- #1704 NCN-702, NCN-440: `com_newsletter` (The current PR: update verification, after porting functionality from custom com_reply)

## Summary

Move and adapt custom code from Nanohub (`com_reply`) into the core CMS component, `com_newsletter`.

Then, update access validation for one-click links to use the secrets (hub, user, and campaign) created in the Epic mentioned above, NCN-434.

NOTE that all PRs mentioned above _are already merged_ into this branch. All of them have been reviewed and tested.

Jira cards (for this PR):
- [NCN-440](https://sdx-sdsc.atlassian.net/NCN-440) Update validation for one-click access 
- [NCN-702](https://sdx-sdsc.atlassian.net/NCN-702) Move com_reply functionality into CMS core com_newsletter
- [NCN-733](https://sdx-sdsc.atlassian.net/NCN-733) Add CSRF tokens and checks

## What's the point again?

Nanohub team wants to be able to send emails to their users with secure, unique links that enable users one-click access to their editable profile details or to a survey. This work ensures that we can selectively reset or expire codes for individual users or for a campaign. In order to provide access we maintain unique secret codes for each user; for each "campaign"; and for the Hub. The three of these must be hashed together and the expiration date validated for access to be granted.

If this solution seems weirdly incomplete, it is. The Nanohub team uses a Salesforce instance to do their mailings. This requires only that they hit a replica of the Nanohub database to query for the needed campaign id, username, and hashed access code. Those are pulled into Salesforce, which handles the mailing part.

However, it was thought that this functionality might be generally useful so this work moves it into core.

Previous PRs provide admin user interfaces for Hub-side management of:
- individual user `secret` codes 
- creation of user secrets for new members
- management of new campaigns (which are secrets with expiration dates)
- management of Hub secret

## Use 

The one-click access code is generated by a call to stored procedure `hash_access_code` as follows:

`SELECT hash_access_code(campaignId, username);`

Here's a partial example of what that looks like:

![the-hash](https://github.com/hubzero/hubzero-cms/assets/3996484/5fe63b59-5bb9-428a-beef-cca56245dc0e)


### Page access

Example format of the secure page-access URLs provided to users:

`https://hub.hubzero.org/newsletter/pages/2?campaign=CAMPAIGNID&user=USERNAME&code=CODE_FROM_hash_access_code`

### Subscription access

`https://hub.hubzero.org/newsletter/email-subscriptions?campaign=CAMPAIGNID&user=USERNAME&code=CODE_FROM_hash_access_code`

Management of the one-click pages, subscriptions, and related tables is no different than it was on nanoHUB, except that any hardcoding of the Hub name was replaced. Management of pages is largely manual and no change to that was scoped in this work (see README section below).

This is an example of the one-click "email subscription" page following successful validation:

![example-one-click-page](https://github.com/hubzero/hubzero-cms/assets/3996484/8a7ee7fc-6f37-4bb7-b421-651dbcb8d796)

## Testing

This work was tested on a developer AWS instance. At the stage of this PR, testing concentrated on behavior when different URLs were provided. In particular:

- sensible messages on failure to validate
- sensible messages on success
- failure to validate on encountering an expired campaign or nonexistent user
- navigation to correct page
- submission of correct page or emailsubscription and update of database entry
- no privileged access of other pages or Hub locations is enabled

Generation of codes via the stored procedure was also tested.

## The README

The following is the updated README taken from the custom `com_reply` component. I reproduce it here to provide context on the work that was done.

## Reply functionality: single-click validation

### Use Case

"Reply" functionality allows users to submit survey responses or update their preferences without first completing the primary hub authentication procedure.

This feature provides user authentication via hashes of randomly-generated codes that are associated with the user's CMS record, the campaign, and the Hub.

This functionality is now part of the `com_newsletter` core component. It was originally found in custom component `com_reply`.

Related features:

All management of these features is found under the Hub's administrator interface.

* Per-user randomly generated codes are managed under Member password management.
* Per-campaign randomly generated codes are managed under the Newsletter component's Campaign tab, on the administrative interface.
* The Hub's randomly generated code is managed under Global Configuration on the administrative interface.

### Install
1. Run the migrations e.g. via `muse migration run ...`

### Adding a Page
It is assumed that all users providing correct credentials will have access to all pages.

To add a form-based page:
1. Add a record to `jos_reply_pages`
1. Create a view for the page in `site/views/pages` 

### Adding an Email Subscription Option
1. Create or confirm name of the corresponding profile field
1. Add a record to `jos_email_subscriptions`
1. Create a view (content to appear beneath label) if desired 
1. Set the email page ID in `secrets/code.php`, using `secrets/code_example.php` as a model.
    * The page ID key is referenced in CodeHelper.php `validateEmailSubscriptionsCode()`.

### Gotchas
* `jos_email_subscriptions.profile_field_name` must correspond with a `jos_user_profile_fields.name`
* `jos_user_profile_options` must be populated for the user profile fields. `jos_user_profile_options.field_id` must correspond with id from `jos_user_profile_fields.id`
* `jos_users` user codes are provided at login time by the core Hubzero members plugin.
* Hashed access codes are provided by the stored procedure, which can be accessed from the database.
    It takes two arguments, `campaign`.`id` and `jos_users`.`username`

### Component-Specific Tables
* `jos_reply_pages` - pages for collecting data
* `jos_reply_replies` - user-submitted responses
* `jos_email_subscriptions` - email subscription options

### Access-Related Tables
* `jos_users` - contains a `secret` for each user
* `campaign` - contains a `secret` for each campaign, with expiration date
* `jos_config` - the `value` column for `scope`='hub' and `key`='secret' contains the overall secret for the Hub

### Integrations
The email subscription update functionality integrates with the core user profile data.
* `jos_user_profile_fields` - profile fields
* `jos_user_profile_options` - profile field response options
* `jos_user_profiles` - associates user data with profile fields

### Generating Access Codes

The stored procedure `hash_access_code` is used to generate and verify codes for page access. Codes are used in URLs as shown below. 

### Example URLs

Here:

* USERNAME is the user's jos_user.username
* CAMPAIGNID is campaign.id

#### pages

Format of links for page access:

`https://jsperhac.aws.hubzero.org/newsletter/pages/2?campaign=CAMPAIGNID&user=USERNAME&code=CODE_FROM_hash_access_code`

#### subscriptions

Format of links for so-called subscription access:

`https://jsperhac.aws.hubzero.org/newsletter/email-subscriptions?campaign=CAMPAIGNID&user=USERNAME&code=CODE_FROM_hash_access_code`

jsperhac committed:
[NCN-438] Create Hub Secret management in com_config

jsperhac committed:
[NCN-437] Create com_newsletter UI for managing Campaign and its secret

jsperhac committed:
[NCN-439] Enable per-user secret management in com_members

jsperhac committed:
[NCN-633] Provide user secret plugin functionality in PlgUserHubzero

jsperhac committed:
[NCN-440] Move and update one-click page access into CMS core

dbenham commented on core/components/com_newsletter/helpers/codeHelper.php:
```diff
@@ -0,0 +1,52 @@
+		// Calculate and compare hash of hub, user, and campaign secrets to passed code:
+		$database = \App::get('db');
+		$user =  $database->quote($username);
+		$sql = "SELECT hash_access_code($campaignId, $user)";
```
Secure this sql just a little bit

jsperhac replied:
Commit 9e598501 has been added; it binds the params to the query prior to execution.

jsperhac committed:
[NCN-440] Bind params for stored procedure call

jsperhac committed:
[NCN-440] Correct routing for the one-click page

jsperhac committed:
[NCN-440] Correct routing for emailsubscriptions, fix hubname display

nkissebe commented on core/components/com_config/admin/controllers/application.php:
```diff
@@ -206,6 +214,35 @@ public function cancelTask()
+		// If user has opted to reset the Hub secret:
+		if ($hubsecret['reset_secret'] == 'reset')
+		{
+			$tableName = 'jos_config';
```
This should be more correctly "#__config" but we have never done anything about changing our table prefixes so while it should be fixed its a "meh"

nkissebe commented on core/components/com_members/migrations/Migration20231016000000ComMembers.php:
```diff
@@ -0,0 +1,109 @@
+		$targetUsers = "UPDATE $tableName
+			SET $column = MD5(CONCAT(NOW(), RAND(), UUID()))
+			WHERE $column IS NULL
+			AND lastVisitDate > DATE_SUB(CURDATE(), INTERVAL 1 YEAR)";
```
I wouldn't have put the date limit. But its fine.

nkissebe commented on core/components/com_newsletter/admin/views/campaigns/tmpl/display.php:
```diff
@@ -0,0 +1,128 @@
+				<th scope="col"><?php echo Html::grid('sort', 'COM_NEWSLETTER_CAMPAIGN', 'title', @$this->filters['sort_Dir'],
+					@$this->filters['sort']);?></th>
+				<th scope="col" class="priority-3"><?php echo Html::grid('sort', 'COM_NEWSLETTER_CAMPAIGN_EXPIRE_DATE', 'expire_date',
+					@$this->filters['sort_Dir'], @$this->filters['sort']); ?></th>
```
I'm not a big fan of using the error supression operator for this purpose @, lets try to avoid in future but not going to block because of it. It hides legit errors you would want to know about, is slow (from a micro optimization viewpoint), and just has a bad feel to it. With PHP 7+ I think there are some equally succinct to handle this moving forward.

nkissebe commented on core/components/com_newsletter/admin/views/campaigns/tmpl/display.php:
```diff
@@ -0,0 +1,128 @@
+							</a>
+						</td>
+						<td class="priority-3">
+							<?php echo Date::of($campaign->expire_date)->toLocal("Y-m-d H:ia"); ?>
```
I think we have a constant defined somewhere for this date string. Not 100% sure. Again not a deal breaker.

jsperhac replied:
Left this as written (and as tested!)

nkissebe commented on core/components/com_newsletter/helpers/subscriptionsHelper.php:
```diff
@@ -0,0 +1,186 @@
+	{
+		$activeRef = self::$activeReferenceField;
+
+		$this->db->setQuery("
```
Another case of not using #__ instead of jos_ for prefix handling. Again non-critical. Repeats through file.

jsperhac replied:
This was brought over directly from `com_reply` and I opted not to make changes at that time.

nkissebe commented on core/components/com_newsletter/migrations/Migration20200731133740AlterEmailSubscriptionsTable.php:
```diff
@@ -0,0 +1,54 @@
+		$now = Date::toSql();
+
+		$renameDescription = "ALTER TABLE $tableName
+													CHANGE COLUMN description profile_field_name varchar(255);";
```
Whitespace is a little funny here. non-blocking

jsperhac replied:
This migration file needs to be renamed anyway so I'll take care of this too. Thanks!

nkissebe commented on core/components/com_newsletter/migrations/Migration20200731133740AlterEmailSubscriptionsTable.php:
```diff
@@ -0,0 +1,54 @@
+		$now = Date::toSql();
+
+		$renameDescription = "ALTER TABLE $tableName
+													CHANGE COLUMN profile_field_name description varchar(255);";
```
whitespace funny here too. non-blocking

jsperhac replied:
This migration file needs to be renamed anyway so I'll take care of this too. Thanks!

nkissebe commented on core/components/com_newsletter/migrations/Migration20231222155500CreateHashStoredProc.php:
```diff
@@ -0,0 +1,36 @@
+class Migration20231222155500CreateHashStoredProc extends Base
+{
+
+	public function up()
```
I wonder if db layer would correctly handle #__ in these.... non-blocking

jsperhac replied:
Again, since I was updating the file I have also handled generalizing the prefix of those table names.

jsperhac committed:
[NCN-440] Correct table name to 'jos_campaign' from 'campaign'

jsperhac committed:
[NCN-440] Rename migration files moved from com_reply component

jsperhac committed:
[NCN-440] Remove migration files and references to unused tables

jsperhac committed:
[NCN-440] Display campaign id in admin campaign manage/edit UI

jsperhac commented:
I've addressed your blocking change requests:
- updated table name from 'campaign' to 'jos_campaign' (commit 57035c7)
- renamed any `com_reply` migration files that had dates prior to 2023 in their names (commit b1b0538)

While doing this I saw a couple of things that needed to be addressed: 
1. Three migration files create and/or modify tables that are not in use in the code. These have been removed and the README.md (documentation) updated. (commit b850c84)
2. Display campaign id in admin UI, so admin user can easily determine it (commit acf60c7)

I've tested all this stuff on my AWS instance.
